### ISSUE TITLE
Use Go-style return values for gocv functions that have result Mat

### ISSUE THREAD

**Issue by willdurand** [hybridgroup/gocv#137]
When reviewing #135, I was like "wait, why is this test actually passing?". The test was about flipping a `Mat` into a brand new `Mat`, both passed to a function that looks roughly like this:

```go
func Flip(src Mat, dst Mat, flipCode int) {}
```

In Go, parameters passed by value (default like above) should not be modified (and in any case, it will never change the value in the parent block). In gocv, this would works though because we manipulate a C pointer under the hood but this is a bit misleading.

Question is: could we agree on a "convention" stating that all arguments modified should be passed by pointer?

If we take the `Flip()` function as an example:

```go
src := ...
dst := gocv.NewMat()

// wait, how does `dst` can be updated? No return-value either? Hmmm.
gocv.Flip(src, dst, 0);

// versus

// oh right so `Flip` takes a `src` Mat and puts the result in `dst`. Lovely <3
gocv.Flip(src, &dst, 0);
```

**Comment by berak** [comment]
somehow, you've hit a spot on the general concept here.

the gocv wrappers all go like this:

    type someclass {
         p pointer_to_some_CGO_implementation
    }
  
so, actually, whatever p points to, will get changed, not the go instance.
(and taking the address of it, won't mend it)

imho, it's already a problem in the underlying opencv c++ code, [for further midbending fun, have a look here]( https://github.com/opencv/opencv/blob/master/modules/core/include/opencv2/core/mat.hpp#L152-L421)

**Comment by willdurand** [comment]
> somehow, returning dst, instead of taking it by a (invisible) reference, like it is done in python, would be a better idea. (too late, too change it for all of the core bindings ?)

I think returning values would be indeed better.

**Comment by berak** [comment]
right, yet it still *works*, like it is, now

**Comment by deadprogram** [comment]
There was a discussion about this at some point on slack or something like that. The question is, should we be more like the C++ API (the current way) or more like the Python API (return a new Mat in function that has a 'dest' that the dev needs to close).

I am still not pulled either way with great conviction, the C++ style seemed the best at the time, but we could decide to change the interface. Seems like doing so sooner would be best if we are going to do so.

**Comment by deadprogram** [comment]
As I recall it was @221bytes that I was discussing this with on the Golang Slack channel.

In any case, we can decide to follow one of 2 approaches here:

- continue using the `dst` Mat assigned pattern as currently implemented. It does make it more like the C++ code, so the C++ docs would match up, but is a little different than the typical Golang code.
- switch to returning a new `gocv.Mat` for any returned values, more like how the Python code handles. It makes it a little less of a match to the C++ docs, but is certainly a little more like what go programmers expect.

I don't like to break the API, but this conversation keeps coming up for good reason. Seems like it would be better to be a little more idiomatic for Go.

To use the example above:

```
src := ...

// dst is now return-value is all cases that there is a Mat that is modified.
dst := gocv.Flip(src, 0);
```

There is, however, a use case that is handled by OpenCV's C++ interface that this would not accommodate, which is when the `src` and `dst` are the same, what openCV calls "inplace". There are a number of OpenCV functions which allow this, for example `dilate`. This is rather inconsistent, as only some OpenCV functions support this. 

I am starting to get the feeling we should probably just ignore the inplace return assignment option and always return new Mat values from any functions that have an output. This would eliminate an entire category for potential confusion for the consumers of this package.

Ideally, I would like to hear from a few more community members before revamping the current API. But so far everyone has been in favor of changing to a more idiomatic Go approach of returning new values.

**Comment by willdurand** [comment]
> continue using the dst Mat assigned pattern as currently implemented. It does make it more like the C++ code, so the C++ docs would match up, but is a little different than the typical Golang code.

I'd be OK with this if we could add pointers for arguments that are modified.

> I don't like to break the API, but this conversation keeps coming up for good reason. Seems like it would be better to be a little more idiomatic for Go.

The good thing is `1.0.0` has not been released yet, so we can still BC break for good reasons.

> I am starting to get the feeling we should probably just ignore the inplace return assignment option and always return new Mat values from any functions that have an output. This would eliminate an entire category for potential confusion for the consumers of this package.

I like this proposal a lot! I believe a package like gocv should provide a uniform/Go-oriented API on top of opencv. The current package is already good but this suggestion would make it even better.

**Comment by deadprogram** [comment]
I changed the title of this to match the change I am proposing. @berak @221bytes anyone else have any thoughts on this change?

**Comment by 221bytes** [comment]
I remember this talk, you said it was easier to stick to the C++ API for documentation and ease of implementation. Indeed, as a total newbie, it was for me easy to contribute (very small contribution though) to the project. 
Except the reason, we wanna do it the gopher way, what are the other pros of refactoring now?

**Comment by willdurand** [comment]
> Except the reason, we wanna do it the gopher way, what are the other pros of refactoring now?

see my initial issue description, mutation happens by accident.

**Comment by 221bytes** [comment]
By accident ? I can't understand what you mean sorry

**Comment by willdurand** [comment]
> By accident ? I can't understand what you mean sorry

Go passes variables by value (copy) not by reference to functions. In Gocv, we should not expect `dst` to be modified after having called `gocv.Flip(src, dst, 0)` since `dst` is passed by value. Yet, it works because `Mat` only attribute is a C pointer. This is very misleading.

**Comment by 221bytes** [comment]
Hum, the name his dest and every functions as a nice comment associated with it. I can't see how someone can be misguided to be honest. I am not a super user and it was clear for me. Refactoring all the library only for this purpose sounds a bit overkill to me. 
Ron took this decision at the beginning and it's working for us. Sticking to the c++ api is nice for documenting and for moving code from c++ to golang so for me we should keep the way it.
Please remember that I was a supporter of a more golang style at first ^^

**Comment by deadprogram** [comment]
Here is a basic example that shows that any approach has its perils.

The current approach reuses the same memory allocated for the `Mat` without the Golang developer having to worry about releasing the OpenCV Mat object underneath:

```go
webcam, err := gocv.VideoCaptureDevice(int(deviceID))
img := gocv.NewMat()

for {
    ok := webcam.Read(img);
}
```

My proposed approach would require a release of the OpenCV Mat object underneath at the end of each loop iteration, or risk running our of memory:

```go
webcam, err := gocv.VideoCaptureDevice(int(deviceID))

for {
    img, ok := webcam.Read()
    ...
    img.Close()
}
```

I think the proposal being made by @willdurand is to do this:

```go
webcam, err := gocv.VideoCaptureDevice(int(deviceID))
img := gocv.NewMat()

for {
    ok := webcam.Read(&img);
}
```

This would still have the implementation just using the C pointer underneath as it is now, but makes code look more like Go programmers expect it to work, if I have it correct.

We certainly cannot avoid the reality that, as Rob Pike put it "CGo is not Go". We can certainly do our best to be as Go-like as possible.

More thoughts/feelings/opinions from the world-at-large?

**Comment by deadprogram** [comment]
I'd love to hear your opinions on this @berak @maruel @drnic or anyone else!

**Comment by maruel** [comment]
Hi! Thanks for the call for opinions. I have opinions! :)

Since I was paged in, I'll state my opinion. On the other hand I haven't used the API in any significant way so I'm probably not the best person to weight in, so I'll state in from the prospect of an API designer, which I happen to have experience in.

I use two goals when designing an API:

1. Reading a snippet of code should make things obvious to an unfamiliar reader, without having to resort to documentation.
2. It shall be designed for one of speed, simplicity of use or consistency with an external API. Know what you optimize for first.

About 1), nobody reads the doc unless they get stuck, you don't want users to get stuck. They'll prefer to copy paste a snippet and start from there. Having an input being mutated in Go does happen in some cases, io.Reader is a notorious example as @deadprogram noted above. In practice, it is important to note that the method Read() accepts a slice, not an struct instance. What's confusing here is that Mat is an struct instance, not an interface. Take as an example io.Copy(), what would be io.Copy(&bytes.Buffer{}, ...) is simply Read(gocv.Mat{}), a pass by value instead of a pointer. I think it's the primary source of confusion.

About 2), it really depends. People in the thread, including Ron himself, noted the value of sticking as close as possible to the C API for documentation purposes. If you want to aim for speed, you want to reduce heap allocations, so you want to accept outputs as arguments, instead of returning them (like io.Reader). If you aim for simplicity, returning a new gocov.Mat instance is likely the most idiomatic way. (I personally value speed, otherwise why using cgo at all?)

Then there's the trade off of breaking the API. In my projects, I "fixed" it by immediately releasing major version 1 and being quite open to releasing new major versions, around once per year or so. I try to batch the breaking API changes into one short amount of time (a ~3 months window), and each breakage is done for a real purpose that users can understand why.

I'm in the camp that trying to be "as close to the C API" when writing a C API wrapper in Go is counter-productive, but the cost of breaking current gocv users may not make doing structural changes like the one I describe below be worth it.

I'll still give my opinion since I was asked about it but feel free to dump it to /dev/null. :)

Examples of things I'd change:

- I'd make a MatFloat64, MatFloat32, MatInt32, MatUint16 (and whatever other types) so that the methods GetXXX() would not be a mess. I'm using image.Image.At() as an example here; https://golang.org/pkg/image/#Image. *Then*, Mat could be an interface! You get strong typing for free. As a matter of fact, it would enable you to make the MatXX to implement https://golang.org/pkg/image/draw/#Drawer, which then would make it possible to decode a jpeg into a gocv.Mat or things like that. 
- Make most functions that "returns" a Mat be methods of Mat, so that `AbsDiff(src1, src2, dst Mat)` would become a method `Mat.AbsDiff(src1, src2 Mat)`. Same for stuff like PutText(). That would make it really more obvious IMHO. Right now https://godoc.org/gocv.io/x/gocv is a mess (sorry) and bucketting the functions into methods would make the API tremendously more readable. Example: https://golang.org/pkg/time/
- I'd change bool return values to be error, in particular Read().
- I'd move UI stuff (Window) into a subpackage gocvui to make it clear it's for ui, not algorithm.
- Same for I/O (IMRead/IMWrite/VideoCapture) into gocvio.
- Remove the implicit dependency on the C API documentation. Make yours be self-standing and be _free_! :)

Then I think the API would be much more idiomatic, more welcoming to new users by being more focused, at the cost of being less C-like.

Is it worth? Is it desirable? I can't make the call but I it's worth taking a deep look and experimenting on this.

**Comment by deadprogram** [comment]
Thanks for the comments @maruel however some of these items are counter the design of not trying to re-architect the internals of OpenCV itself. For example, the UI stuff is in the OpenCV `highgui` module in the main package, so as to allow people to find where to add things within GoCV. I think we discussed this via email early on.

In any case, lots of ideas in your thoughtful response. So there is a possible new option. To restate all the possibilities:

### Current code:
```go
dst := gocv.NewMat()
gocv.Flip(src, dst, 0);
```

### Pass dst as a pointer to avoid "surprise" to Go programmers:
```go
dst := gocv.NewMat()
gocv.Flip(src, &dst, 0);
```

### Return a new Mat as dst:
```go
dst := gocv.Flip(src, 0);
```

### Change functions that mutate dst into methods off of the Mat they change:
```go
dst := gocv.NewMat()
dst.Flip(src, 0);
```

**Comment by deadprogram** [comment]
Another use case to consider is when using any algorithm's `Apply()` function.

### Current code

```go
src := ...
dst := gocv.NewMat()

mog2 := gocv.NewBackgroundSubtractorMOG2()
mog2.Apply(src, dst)
```

### Pass dst as a pointer to avoid "surprise" to Go programmers

```go
src := ...
dst := gocv.NewMat()

mog2 := gocv.NewBackgroundSubtractorMOG2()
mog2.Apply(src, &dst)
```

### Return a new Mat as dst

```go
src := ...

mog2 := gocv.NewBackgroundSubtractorMOG2()
dst := mog2.Apply(src)
```

### Change functions that mutate dst into methods off of the Mat they change:

```go
// not sure how to use this pattern here...
```

Seems like the mismatch with option 4 in this use case is problematic.

**Comment by deadprogram** [comment]
At the moment, I'm starting to get the feeling that the @willdurand original suggestion of passing `Mat*` to indicate output param might be the best balance between Go-like idiom (anything passed by pointer is subject to mutation) and performance (we do not need to allocate/de-allocate extra cv::Mat objects within loops).

This would also probably be the smallest change to the API needed for consumers of the package (adding `&` in front of any `dst` params) vs. the other proposed changes.

**Comment by maruel** [comment]
Makes sense given the trade offs.

(Only for completeness)

The form would look like:

### Change functions that mutate dst into methods off of the Mat they change:

```
dst := gocv.NewMat()
mog2 := gocv.NewBackgroundSubtractorMOG2()
dst.Apply(mog2)
```

**Comment by deadprogram** [comment]
Actually I think that would be

### Change functions that mutate dst into methods off of the Mat they change:

```go
src := ...
dst := gocv.NewMat()
mog2 := gocv.NewBackgroundSubtractorMOG2()
dst.Apply(mog2, src)
```

**Comment by deadprogram** [comment]
This is my WIP branch implementing the `Mat*` approach:

https://github.com/hybridgroup/gocv/tree/feature/mat-pointer

**Comment by maruel** [comment]
I think it definitely helps, especially for the cases where dst and src are reversed.

**Comment by willdurand** [comment]
Thanks for working on this, let me know if you need help, I have some spare time for this lib.

I skipped @maruel's comment about documentation, I also think that it would be better for end users to have the documentation as part of the godoc, not links to opencv documentation. What if the urls change for some reason?

**Comment by deadprogram** [comment]
The OpenCV docs have hashes that do not change between versions, and now that we adopted earlier @berak suggestion as discussed in https://github.com/hybridgroup/gocv/issues/109 to just link to master branch version, the links will always be the latest.

I like linking to the OpenCV docs as it gives us the full benefit of the OpenCV community's docs. But I do not like *requiring* devs to look at both places to do anything useful. 

We ideally should have "good enuf'" GoDocs that do not require always going to the OpenCV docs. That will require a lot more effort on our parts to do so but for sure will be worth it. To achieve this we need to include a little more info for each function, what would the ideal function 

Regarding the @maruel comment on moving things to be methods of `Mat`. The problem is that since nearly everything uses `Mat` it would not really help much as far as "unflattening" the docs. I do not want to use sub-packages that do not match the OpenCV modules layout for reasons already expressed. Hence our options are limited on that.

**Comment by willdurand** [comment]
Roger that, I forgot about the links.

> I do not want to use sub-packages that do not match the OpenCV modules layout for reasons already expressed. Hence our options are limited on that.

The highgui stuff could be put in a sub-package, maybe? It is part of the opencv "highgui" module. WDYT?

**Comment by deadprogram** [comment]
In addition to helping newcomers to this project but who know OpenCV, or people who need some specific unimplemented functionality, the current structure mapping directly to OpenCV's modules works well. Are those the designs patterns we might otherwwise choose? IDK but OpenCV is the target, so we want to reduce any cognitive load needed to find things. 

Also, CGo does not work well with subpackages in all my previous experiments.

**Comment by deadprogram** [comment]
I am very close to merging the https://github.com/hybridgroup/gocv/tree/feature/mat-pointer branch into the `dev` branch, and committing to this style of API for the 1.x GoCV.

Any last comments/feedback/critiques/requests before this becomes "official"?

**Comment by willdurand** [comment]
do you have a PR opened?

**Comment by deadprogram** [comment]
Here it is https://github.com/hybridgroup/gocv/pull/142

**Comment by deadprogram** [comment]
This has now been released as part of v0.11.0 thanks again everyone! Now closing.
